### PR TITLE
Add runtime option to use DM that is 'similar' to the mother grids in the PML.

### PR DESCRIPTION
**Pull request by atmyers**
This takes advantage of this new function in amrex: https://github.com/AMReX-Codes/amrex/pull/2526

- [x] Check performance numbers when the new option is turned on/off.
- [ ] Do a few more checks to understand why the checksum is modified by this option.

### CODE REVIEW

**Comment by RemiLehe** [comment]
Thanks! Do you understand why some automated tests are failing?

**Comment by atmyers** [comment]
Note that this PR currently adds the option and also turns it on to test the effect of doing so.

The following tests are currently failing:
- Python_wrappers (crashes)
- PEC_Particle (3D)
- pml_psatd_dive_divb_cleaning (3D)
- Repelling particles (2D)
- subcyclingMR (2D)

"SubcyclingMR", "PEC_Particle" and "Repelling particles" I think are clearly roundoff / particle order issues that could be affected by the dmaps being different. "pml_psatd_dive_divb_cleaning" could potentially be roundoff but the diffs maybe look a little larger than I would expect - maybe you could take a look and see what you think? 

I do not know why the Python_wrapper test crashes at the moment.

**Comment by ax3l** [comment]
Let's rebase this one once #2666 is merged

**Comment by ax3l** [comment]
@atmyers ready for rebase :)

**Comment by ax3l** [comment]
Curious:
- `Python_wrappers CRASHED (script failed)`
- `pml_psatd_dive_divb_cleaning FAILED` (checksum)

**Comment by ax3l** [comment]
Let's see if a rebase brings clarity. Maybe the merged `development` was not up-to-date.

File: Source/WarpX.cpp [hunk 1]
```diff
@@ -744,6 +746,7 @@ WarpX::ReadParameters ()
         pp_warpx.query("pml_has_particles", pml_has_particles);
         pp_warpx.query("do_pml_j_damping", do_pml_j_damping);
         pp_warpx.query("do_pml_in_domain", do_pml_in_domain);
+        pp_warpx.query("do_similar_dm_pml", do_similar_dm_pml);
```
Comment (ax3l) [inline Source/WarpX.cpp:749]: Please add docs in `Docs/source/usage/parameters.rst` :)

**Comment by EZoni** [comment]
The test `Python_wrappers` passes if you change the relative tolerance set here
https://github.com/ECP-WarpX/WarpX/blob/603f75d295f5d87baf3284807319e027a4b93a1c/Examples/Tests/PythonWrappers/PICMI_inputs_2d.py#L251-L252
from `1e-09` to `2e-09`. The reason is that we have some hard-coded benchmarks in the PICMI script (similar to our usual checksum benchmarks but computed directly from the Python data) and the data computed from the third component of Ez in the PML has a relative deviation from the benchmark that happens to be slightly larger than `1e-09`, probably due to the changes in this PR. This is the benchmark value I'm talking about:
https://github.com/ECP-WarpX/WarpX/blob/603f75d295f5d87baf3284807319e027a4b93a1c/Examples/Tests/PythonWrappers/PICMI_inputs_2d.py#L274
and the value produced (locally, at least) by the test is now `190919663.47672307` instead, which passes the check if `rtol = 2e-09`.

Since this change seems reasonable, we have two options: (i) increase the relative tolerance `rtol` from `1e-09` to `2e-09`, or (ii) reset by hand the benchmark value using the new value I just posted and leave the relative tolerance `rtol` unchanged.

Regardless of whether we choose option (i) or (ii) to fix the test for this PR, in a separate PR we could maybe try to make that part of the PICMI script more user-friendly when it comes to crashes like this one. Do you have any suggestions, @ax3l or @atmyers?

**Comment by atmyers** [comment]
Ah, thank you for finding that! I tend to think we should bump the tolerance if it's tight enough that changing the parallel layout of the PML data makes it fail.

The `regtest.py` script writes errors generated by the WarpX python scripts to a file for later inspection, but that is lost when running through CI - perhaps we could modify that to optionally also show this information to stdout when running the test suite?

**Comment by EZoni** [comment]
Regarding the test `pml_psatd_dive_divb_cleaning`, I agree that the relative changes in the checksum benchmarks are slightly larger than one might expect (I see errors of order `1e-06` overall, locally), but maybe that is due to the fact that the test uses PML boundary conditions in each direction in 3D, so maybe the round-off errors in each direction kind of combine together when looking at the effects on the fields in the regular grid. Do you think this could make sense? Consistently with the errors in the benchmarks (which one cannot spot by eye), the plots look similar between the results produced on `development` and the results produced on this branch.

**Comment by atmyers** [comment]
It does seem like this makes a significant improvement to the runtime for runs with PMLs. 

Total runtime and top 5 routines by exclusive time, with `warpx.do_similar_dm_pml=1`:
```
TinyProfiler total time across processes [min...avg...max]: 75.01 ... 75.01 ... 75.01

----------------------------------------------------------------------------------------------------------
Name                                                       NCalls  Excl. Min  Excl. Avg  Excl. Max   Max %
----------------------------------------------------------------------------------------------------------
FillBoundary_finish()                                       44871      8.299      18.28      24.65  32.86%
FabArray::ParallelCopy_finish()                             54012      7.226      12.41      20.37  27.16%
FillBoundary_nowait()                                       44871      4.391      6.247       8.49  11.32%
Filter::ApplyStencil(FArrayBox)                             21074       1.06       4.86      8.303  11.07%
FabArray::ParallelCopy_nowait()                             54012      3.706      4.399      6.199   8.26%
```

Same thing, but with `warpx.do_similar_dm_pml=0`:
```
TinyProfiler total time across processes [min...avg...max]: 104.7 ... 104.7 ... 104.7

----------------------------------------------------------------------------------------------------------
Name                                                       NCalls  Excl. Min  Excl. Avg  Excl. Max   Max %
----------------------------------------------------------------------------------------------------------
FabArray::ParallelCopy_finish()                             54012      7.516      32.65      53.67  51.28%
FillBoundary_finish()                                       44871      5.149      25.55      49.88  47.66%
FillBoundary_nowait()                                       44871      3.862      7.037      11.67  11.15%
Filter::ApplyStencil(FArrayBox)                             21074      1.176      4.965      8.282   7.91%
FabArray::ParallelCopy_nowait()                             54012      3.777      5.424      8.027   7.67%
```

The inputs file was basically `Examples/Physics_applications/plasma_acceleration/inputs_3d_boost`, with the number of steps reduced and IO turned off.

**Comment by ax3l** [comment]
That's fantastic news, thanks a lot! :tada: 

For the checksum change in `pml_psatd_dive_divb_cleaning` we should check manually with a plot to be sure we see no new artifacts - @EZoni can you help with that please? :)

**Comment by EZoni** [comment]
@ax3l @RemiLehe @atmyers 
Regarding the test `pml_psatd_dive_divb_cleaning`, I've done a few tests locally as we discussed last week (with Remi and Andrew) and I have two observations:
- the errors in the checksum benchmarks remain of order `1e-06` even if I turn off div(E) and div(B) cleaning in the regular domain;
- looking at some plots (e.g., for Ez on the y=0 plane) of the difference between the data with and without the new option, there doesn't seem to be a specific region in the domain where the largest errors are localized, it seems pretty much spread out.

My suggestion at this point, in order to merge this PR quickly, would be to set `warpx.do_similar_dm_pml = 0` for the test `pml_psatd_dive_divb_cleaning` so that we don't change its benchmarks and we can try to look into this further. What do you think?

File: Docs/source/usage/parameters.rst [hunk 2]
```diff
@@ -277,6 +277,10 @@ Additional PML parameters
 * ``warpx.pml_ncell`` (`int`; default: 10)
     The depth of the PML, in number of cells.
 
+* ``do_similar_dm_pml`` (`int`; default: 0)
```
Comment (EZoni) [inline Docs/source/usage/parameters.rst]: The new default seems to be `1`, right?
```suggestion
* ``do_similar_dm_pml`` (`int`; default: 1)
```

**Review by RemiLehe — APPROVED** [review]

**Pull request merged**